### ISSUE TITLE
Resume after interrupted `pgcopydb follow` throws "Failed to parse BEGIN: transaction already in progress"

### ISSUE THREAD

**Issue by arajkumar** [dimitri/pgcopydb#471]
I see the following error when I resume the interrupted `pgcopydb follow`

```
07:51:48.966 2311 INFO   main.c:163                Running pgcopydb version 0.13.21.gf14255a.dirty from "/home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/pgcopydb"
07:51:49.029 2311 ERROR  ld_transform.c:1000       Failed to parse BEGIN: transaction already in progress
07:51:49.029 2311 ERROR  ld_transform.c:827        Failed to parse JSON message: {"action":"B","xid":"3169590","lsn":"9D/757932C0","timestamp":"2023-09-19 05:04:39.159606+0000","message":{"action":"B","xid":3169590}}
```

I drilled into the json file produced by the `pgcopydb receive` process, it has two BEGIN messages for the same transaction. Here is a [trimmed json file](https://github.com/dimitri/pgcopydb/files/12657316/000000020000009D00000075.json.txt) and the problem can be reproduced using `pgcopydb stream transform dir/cdc/000000020000009D00000075.json /tmp/test.sql`

## Steps to reproduce the error

1) Start pgcopydb follow
```
pgcopydb follow --dir $PGCOPY_DIR --plugin wal2json
```
2) `killall -9 pgcopydb` and resume pgcopydb
```
pgcopydb follow --dir $PGCOPY_DIR --plugin wal2json --resume
```

**Comment by arajkumar** [comment]
When I restart after a simulated crash(kill -9), sometime I see errors like below,
```
08:39:31.654 5058 INFO   Resuming streaming from latest file "/home/ubuntu/.local/share/pgcopydb/00000002000000A100000091.json.partial"
08:39:31.710 5058 ERROR  Failed to parse JSON message: {"action":"I","xid":"3274811","lsn":"A1/9119D5A8","timestamp":"2023-09-19 08:38:14.853534+0000","message":{"action":"I","xid":3274811,"schema":"_timescaledb_internal","table":"_hyper_5_458_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-30 04:34:20+00"},{"name":"tags_id","type":"integer","value":344109},{"name":"fuel_state","type":"double precision","value":0.3},{"name":"current_load","type":"double precision","value":2798},
08:39:31.804 5056 ERROR  Subprocess prefetch with pid 5058 has exited with error code 12
08:39:31.804 5059 ERROR  Failed to read stream context file: retry policy expired
08:39:31.804 5060 ERROR  Failed to read stream context file: retry policy expired
08:39:31.804 5060 ERROR  Failed to read the streaming context information from the source database, see above for details
08:39:31.804 5060 ERROR  Failed to setup for catchup, see above for details
08:39:31.954 5056 ERROR  Subprocess catchup with pid 5060 has exited with error code 12
08:39:32.104 5056 ERROR  Some sub-process exited with errors, see above for details
08:39:32.104 5056 ERROR  Failed to follow changes from source, see above for details
```

I usually remove the partial json file(rm ~/.local/share/pgcopydb/*.json.partial
) to keep the `--resume` happy. Is that the right way to do resume the follow?

Few a times resume after deleting the partial file causes this issue.

```
08:42:33.439 5149 INFO   Transformed 68960 JSON messages into SQL file "/home/ubuntu/.local/share/pgcopydb/00000002000000A10000008B.sql"
08:42:33.441 5149 ERROR  Failed to parse BEGIN: transaction already in progress
08:42:33.442 5149 ERROR  Failed to parse JSON message: {"action":"B","xid":"3274134","lsn":"A1/8B5E1168","timestamp":"2023-09-19 08:42:32.916131+0000","message":{"action":"B","xid":3274134}}
08:42:33.442 5149 ERROR  Failed to transform JSON messages from input stream, see above for details
08:42:33.495 5142 ERROR  Subprocess transform with pid 5149 has exited with error code 12
08:42:33.554 5150 INFO   Replayed up to replay_lsn A1/9119BD50, stopping
```

**Comment by arajkumar** [comment]
Probably I should just remove last line from `/home/ubuntu/.local/share/pgcopydb/latest` which is partially written and then resume?

**Comment by dimitri** [comment]
It looks like a bug we want to fix in pgcopydb itself. Can you dive into the topic and see how that second BEGIN message manages to make its way to the JSON file? I believe we have a mechanism in place that prevents that situation from happening already...

**Comment by arajkumar** [comment]
@dimitri Sure, I'm trying to understand [streamCheckResumePosition](https://github.com/dimitri/pgcopydb/blob/4ff2963cdb9f9252f9f15f19309ff036b37ddaf6/src/bin/pgcopydb/ld_stream.c#L633-L637). Hope that is the right place.

**Comment by shubhamdhama** [comment]
While fixing https://github.com/dimitri/pgcopydb/issues/471 I made this assumption that `KEEPALIVE` wouldn't be the last message of an incomplete transaction. It wasn't guaranteed as per the code but handling `KEEPALIVE` was bit more complex so we decided to go incrementally and fix whatever fixed in #471 now and handle KEEPALIVE later.
You might want to look at,
https://github.com/dimitri/pgcopydb/blob/4ff2963cdb9f9252f9f15f19309ff036b37ddaf6/src/bin/pgcopydb/ld_stream.c#L320-L341

@arajkumar I hope this would be useful.

**Comment by arajkumar** [comment]
@shubhamdhama Thanks for the pointers.

>While fixing https://github.com/dimitri/pgcopydb/issues/471 I made this assumption that KEEPALIVE

Did you mean to refer to another issue and not #471?

**Comment by arajkumar** [comment]
What is the recommended method to resume when the latest file ends with a partial json line?

For example, here is the last few lines of "/home/ubuntu/.local/share/pgcopydb/00000002000000B10000009A.json.partial"
```json
{"action":"I","xid":"3738742","lsn":"B1/9AC24398","timestamp":"2023-09-20 06:31:07.342850+0000","message":{"action":"I","xid":3738742,"schema":"_timescaledb_internal","table":"_hyper_6_1213_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-30 10:54:00+00"},{"name":"tags_id","type":"integer","value":878283},{"name":"latitude","type":"double precision","value":5.69364},{"name":"longitude","type":"double precision","value":156.86336},{"name":"elevation","type":"double precision","value":422},{"name":"velocity","type":"double precision","value":35},{"name":"heading","type":"double precision","value":257},{"name":"grade","type":"double precision","value":68},{"name":"fuel_consumption","type":"double precision","value":41.5},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"3738742","lsn":"B1/9AC24398","timestamp":"2023-09-20 06:31:07.342863+0000","message":{"action":"I","xid":3738742,"schema":"_timescaledb_internal","table":"_hyper_6_1213_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-30 10:54:10+00"},{"name":"tags_id","type":"integer","value":873179},{"name":"latitude","type":"double precision","value":81.85788},{"name":"longitude","type":"double precision","value":100.26494},{"name":"elevation","type":"double precision","value":278},{"name":"velocity","type":"double precision","value":24},{"name":"heading","type":"double precision","value":174},{"name":"grade","type":"double precision","value":36},{"name":"fuel_consumption","type":"double precision","value":35.5},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"3738742","lsn":"B1/9AC24398","timestamp":"2023-09-20 06:31:07.342872+0000","message":{"action":"I","xid":3738742,"schema":"_timescaledb_internal","table":"_hyper_6_1213_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-30 10:54:10+00"},{"name":"tags_id","type":"integer","value":878283},{"name":"latitude","type":"double precision","value":5.6955},{"name":"longitude","type":"double precision","value":156.86414},{"name":"elevation","type":"double precision","value":421},{"name":"velocity","type":"double precision","value":40},{"name":"heading","type":"double precision","value":260},{"name":"grade","type":"double precision","value":65},{"name":"fuel_consumption","type":"double precision","value":41.3},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"3738742","lsn":"B1/9AC24398","timestamp":"2023-09-20 06:31:07.342880+0000","message":{"action":"I","xid":3738742,"schema":"_timescaledb_internal","table":"_hyper_6_1213_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-30 10:54
```
As I mentioned, I used to delete the partial file and resume. However, that doesn't seem to work.

**Comment by arajkumar** [comment]
It works all the time when I interrupt the `pgcopydb follow` with Ctrl C & resume. The problem seem to be specific to `killall -9 pgcopydb`.

**Comment by shubhamdhama** [comment]
> @shubhamdhama Thanks for the pointers.
> 
> > While fixing #471 I made this assumption that KEEPALIVE
> 
> Did you mean to refer to another issue and not #471?

Oops. It should be https://github.com/dimitri/pgcopydb/pull/394 (fixed in the orignal message too).

Yes that partial line in `.json.partial` seems problematic. Let's think about the fix here. As we work only in append only mode with respect to these files so removing this line on restart isn't an option. So in my opinion one option is to ignore any invalid JSON lines in these files across the processes (stream, transform, replay). Or atomically write a line. @dimitri what do you think?

**Comment by dimitri** [comment]
In the face of `kill -9` (which is not a signal, the process never receives a signal, it's just removed from the OS list of active processes) ; I'm not sure that it's possible to “atomically write a line” without resorting to a transactional system of sorts. 

That said, in our case I don't suppose the partial JSON line LSN has been reported as written or flushed to the source system, so upon reconnecting we should receive the data again. We might get away with the situation by simply removing the last line of the partial file when we fail to parse it.

**Comment by arajkumar** [comment]
When I look at the resume log, I see that pgcopydb requests to start the replication from `B9/427B6A38`, however the server seem to starts with different LSN which is higher than what we requested `LSN B9/428121A8` and it streams data for `B9/427B6A38` latter.

```
2023-09-21 06:41:25.324 66777 INFO   ld_stream.c:613           Resuming streaming at LSN B9/427B6A38 from first message with that LSN read in JSON file "/home/ubuntu/.local/share/pgcopydb/00000002000000B900000042.json.partial", line 37311
2023-09-21 06:41:25.324 66777 NOTICE ld_stream.c:621           Resume replication from latest message: {"action":"I","xid":"3958454","lsn":"B9/427B5040","timestamp":"2023-09-21 06:25:39.180664+0000","message":{"action":"I","xid":3958454,"schema":"_timescaledb_internal","table":"_hyper_6_1536_chunk","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 07:23:10+00"},{"name":"tags_id","type":"integer","value":1369399},{"name":"latitude","type":"double precision","value":41.95234},{"name":"longitude","type":"double precision","value":136.27363},{"name":"elevation","type":"double precision","value":70},{"name":"velocity","type":"double precision","value":57},{"name":"heading","type":"double precision","value":144},{"name":"grade","type":"double precision","value":100},{"name":"fuel_consumption","type":"double precision","value":41.9},{"name":"additional_tags","type":"jsonb","value":null}]}}
2023-09-21 06:41:25.324 66777 SQL    pgsql.c:508               Connecting to [source] "postgres://tsdbadmin@tsdb-22e8519b-internal-90d0.a.timescaledb.io:26479/defaultdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
2023-09-21 06:41:25.340 66777 SQL    pgsql.c:1493              [SOURCE 2091711] SELECT restart_lsn FROM pg_replication_slots WHERE slot_name = $1;
2023-09-21 06:41:25.340 66777 SQL    pgsql.c:1506              [SOURCE 2091711] 'pgcopydb'
2023-09-21 06:41:25.346 66777 SQL    pgsql.c:418               Disconnecting from [source] "postgres://tsdbadmin@tsdb-22e8519b-internal-90d0.a.timescaledb.io:26479/defaultdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
2023-09-21 06:41:25.347 66777 NOTICE ld_stream.c:407           Connecting to logical decoding replication stream
2023-09-21 06:41:25.347 66777 DEBUG  ld_stream.c:428           startLogicalStreaming: wal2json (6)
2023-09-21 06:41:25.347 66777 SQL    pgsql.c:3736              starting log streaming at B9/427B6A38 (slot pgcopydb)
2023-09-21 06:41:25.347 66777 SQL    pgsql.c:508               Connecting to [source] "postgres://tsdbadmin@tsdb-22e8519b-internal-90d0.a.timescaledb.io:26479/defaultdb?keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60&sslmode=require&replication=database"
2023-09-21 06:41:25.361 66777 SQL    pgsql.c:2896              IDENTIFY_SYSTEM: timeline 2, xlogpos B9/48000188, systemid 7270795386924863497
2023-09-21 06:41:25.362 66777 SQL    pgsql.c:2952              TIMELINE_HISTORY: "00000002.history", timeline 2 started at 45/3E043158
2023-09-21 06:41:25.362 66777 SQL    pgsql.c:4516              RetrieveWalSegSize: 16777216
2023-09-21 06:41:25.362 66777 SQL    pgsql.c:3797              START_REPLICATION SLOT "pgcopydb" LOGICAL B9/427B6A38 ("format-version" '2', "include-xids" 'true', "include-schemas" 'true', "include-transaction" 'true', "include-types" 'true', "filter-tables" 'pgcopydb.*')
2023-09-21 06:41:25.364 66777 SQL    pgsql.c:3812              streaming initiated
2023-09-21 06:41:25.364 66777 DEBUG  ld_stream.c:2536          Wrote wal_segment_size 16777216 into "/home/ubuntu/.local/share/pgcopydb/wal_segment_size"
2023-09-21 06:41:25.364 66777 DEBUG  ld_stream.c:2550          Wrote tli 2 timeline file "/home/ubuntu/.local/share/pgcopydb/tli"
2023-09-21 06:41:25.364 66777 DEBUG  ld_stream.c:2560          Wrote timeline history file "/home/ubuntu/.local/share/pgcopydb/tli.history"
2023-09-21 06:41:25.364 66777 SQL    pgsql.c:508               Connecting to [source] "postgres://tsdbadmin@tsdb-22e8519b-internal-90d0.a.timescaledb.io:26479/defaultdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
2023-09-21 06:41:25.383 66777 SQL    pgsql.c:1493              [SOURCE 2091713] update pgcopydb.sentinel set startpos = $2, write_lsn = $1, flush_lsn = $2 returning startpos, endpos, apply, write_lsn, flush_lsn, replay_lsn;
2023-09-21 06:41:25.383 66777 SQL    pgsql.c:1506              [SOURCE 2091713] 'B9/427B6A38', 'B9/427B6A38'
2023-09-21 06:41:25.394 66777 SQL    pgsql.c:418               Disconnecting from [source] "postgres://tsdbadmin@tsdb-22e8519b-internal-90d0.a.timescaledb.io:26479/defaultdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
2023-09-21 06:41:25.394 66777 DEBUG  ld_stream.c:1472          streamFeedback: written B9/427B6A38 flushed B9/427B6A38 applied 0/0  startpos B9/427B6A38 endpos 0/0 apply disabled
2023-09-21 06:41:25.394 66777 INFO   pgsql.c:4372              Reported write_lsn B9/427B6A38, flush_lsn B9/427B6A38, replay_lsn 0/0
2023-09-21 06:41:26.048 66779 DEBUG  pgsql.c:5270              pgsql_get_sentinel: replay_lsn 0/0
2023-09-21 06:41:26.048 66779 DEBUG  ld_apply.c:330            stream_apply_wait_for_sentinel: previous lsn 0/0, replay_lsn 0/0
2023-09-21 06:41:26.048 66779 DEBUG  ld_apply.c:335            startpos B9/427B6A38 endpos 0/0 apply disabled
2023-09-21 06:41:26.922 66777 DEBUG  ld_stream.c:698           Received action B for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.923 66777 NOTICE ld_stream.c:1085          Found pre-existing partial file "/home/ubuntu/.local/share/pgcopydb/00000002000000B900000042.json.partial"
2023-09-21 06:41:26.923 66777 NOTICE ld_stream.c:1104          Now streaming changes to "/home/ubuntu/.local/share/pgcopydb/00000002000000B900000042.json.partial"
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:1963          stream_update_latest_symlink: "/home/ubuntu/.local/share/pgcopydb/latest" -> "/home/ubuntu/.local/share/pgcopydb/00000002000000B900000042.json.partial"
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/428121A8
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.924 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.925 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
2023-09-21 06:41:26.926 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958459 at LSN B9/42812780
..
..
..
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B5040
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.102 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38
2023-09-21 06:41:27.103 66777 DEBUG  ld_stream.c:698           Received action I for XID 3958454 at LSN B9/427B6A38

```

**Comment by arajkumar** [comment]
I think the issue is not just specific to force kill using SIGTERM. When I quit pgcopydb with Ctrl C and resume, I see the logical messages for already COMMITTED transactions.

```
  65095 {"action":"I","xid":"4527970","lsn":"CC/92F3A558","timestamp":"2023-09-22 06:04:06.3344
  65095 90+0000","message":{"action":"I","xid":4527970,"schema":"_timescaledb_internal","table"
  65095 :"_hyper_5_2484_chunk","columns":[{"name":"time","type":"timestamp with time zone","val
  65095 ue":"2016-01-25 00:07:20+00"},{"name":"tags_id","type":"integer","value":1898677},{"nam
  65095 e":"fuel_state","type":"double precision","value":0.6},{"name":"current_load","type":"d
  65095 ouble precision","value":2450},{"name":"status","type":"double precision","value":2},{"
  65095 name":"additional_tags","type":"jsonb","value":null}]}}
  65096 {"action":"C","xid":"4527970","lsn":"CC/92F3C000","timestamp":"2023-09-22 06:04:06.3344
  65096 94+0000","message":{"action":"C","xid":4527970}}
  ....

  68984 {"action":"I","xid":"4527972","lsn":"CC/92F017E8","timestamp":"2023-09-22 06:04:06.4728
  68984 64+0000","message":{"action":"I","xid":4527972,"schema":"_timescaledb_internal","table"
  68984 :"_hyper_5_2482_chunk","columns":[{"name":"time","type":"timestamp with time zone","val
  68984 ue":"2016-01-24 23:35:20+00"},{"name":"tags_id","type":"integer","value":1889748},{"nam
  68984 e":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"d
  68984 ouble precision","value":2342},{"name":"status","type":"double precision","value":1},{"
  68984 name":"additional_tags","type":"jsonb","value":null}]}}
  68985 {"action":"K","lsn":"CC/92F017E8","timestamp":"2023-09-22 06:04:06.472864+0000"}
  68986 {"action":"I","xid":"4527970","lsn":"CC/92F3A1F0","timestamp":"2023-09-22 06:04:57.2229  68986 33+0000","message":{"action":"I","xid":4527970,"schema":"_timescaledb_internal","table"  68986 :"_hyper_5_2484_chunk","columns":[{"name":"time","type":"timestamp with time zone","val  68986 ue":"2016-01-25 00:00:00+00"},{"name":"tags_id","type":"integer","value":1898677},{"nam  68986 e":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"d  68986 ouble precision","value":null},{"name":"status","type":"double precision","value":5},{"  68986 name":"additional_tags","type":"jsonb","value":null}]}}
...
  69026 {"action":"I","xid":"4527970","lsn":"CC/92F3A558","timestamp":"2023-09-22 06:04:57.2231
  69026 59+0000","message":{"action":"I","xid":4527970,"schema":"_timescaledb_internal","table"
  69026 :"_hyper_5_2484_chunk","columns":[{"name":"time","type":"timestamp with time zone","val
  69026 ue":"2016-01-25 00:07:20+00"},{"name":"tags_id","type":"integer","value":1898677},{"nam
  69026 e":"fuel_state","type":"double precision","value":0.6},{"name":"current_load","type":"d
  69026 ouble precision","value":2450},{"name":"status","type":"double precision","value":2},{"
  69026 name":"additional_tags","type":"jsonb","value":null}]}}
  69027 {"action":"C","xid":"4527970","lsn":"CC/92F3C000","timestamp":"2023-09-22 06:04:57.2231
  69027 64+0000","message":{"action":"C","xid":4527970}}
``` 
When I quite using Ctrl C, `latest` file had contents upto line number 68985, when I resume it starts with Insert messages for a transaction 4527970 which is already completed.

@dimitri Is this is a correct behavior?

**Comment by arajkumar** [comment]
I could reproduce the same problem even with stopping pgcopydb using Ctrl C., So all you need to do is stop the pgcopydb with Ctrl C and then resume it.

```
10:24:51.089 29120 NOTICE Starting the prefetch sub-process
10:24:51.089 29121 NOTICE Starting the transform sub-process
10:24:51.089 29121 NOTICE Transforming from CE/64BFA728 in "/home/ubuntu/.local/share/pgcopydb/00000002000000CE00000064.sql"
10:24:51.090 29121 NOTICE Transforming JSON file "/home/ubuntu/.local/share/pgcopydb/00000002000000CE00000064.json" into SQL file "/home/ubuntu/.local/share/pgcopydb/00000002000000CE00000064.sql"
10:24:51.090 29122 NOTICE Starting the catchup sub-process
10:24:51.090 29120 INFO   Resuming streaming from latest file "/home/ubuntu/.local/share/pgcopydb/00000002000000CE00000064.json"
10:24:51.114 29121 ERROR  Failed to parse BEGIN: transaction already in progress
10:24:51.114 29121 ERROR  Failed to parse JSON message: {"action":"B","xid":"4577122","lsn":"CE/64002428","timestamp":"2023-09-22 10:24:45.861197+0000","message":{"action":"B","xid":4577122}}
10:24:51.116 29121 ERROR  Failed to resume transforming from existing file "/home/ubuntu/.local/share/pgcopydb/00000002000000CE00000064.sql"
10:24:51.116 29121 ERROR  Failed to resume streaming from CE/64BFA728
10:24:51.127 29122 WARN   Failed to connect to "postgres://tsdbadmin@host/tsdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60", retrying until the server is ready
10:24:51.133 29122 SQL    PQping(postgres://tsdbadmin@host/tsdb?sslmode=require&keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60): slept 5 ms on attempt 2
10:24:51.150 29122 SQL    PQping OK after 2 attempts
```

**Comment by dimitri** [comment]
It is expected that resuming logical decoding could send some LSNs that were already processed yes. That's to do with the protocol reporting of writte_lsn, flush_lsn, and apply_lsn. That said, it's easy on the client-side to skip already written LSNs when receiving them again.

What's unexpected is to not be able to receive missing LSN between what we have on-disk and the current replication slot position. If that happens, it's probably a bug in our client for logical decoding in the way we report (too early) LSN progress to update the replication slot.

**Comment by arajkumar** [comment]
@dimitri , if I understand correctly, there are 3 approaches being used to skip already processed LSNs:

1. Get the last processed LSN from the [`latest` file](https://github.com/dimitri/pgcopydb/blob/379d242baca0d0fa9bf358d7566bbe74e466e87c/src/bin/pgcopydb/ld_stream.c#L609).
2. If the `latest` file is empty, use the [startpos from the sentinel table](https://github.com/dimitri/pgcopydb/blob/379d242baca0d0fa9bf358d7566bbe74e466e87c/src/bin/pgcopydb/ld_stream.c#L566C22-L566C40).
3. Use the LSN from the [replication origin progress tracker](https://github.com/dimitri/pgcopydb/blob/379d242baca0d0fa9bf358d7566bbe74e466e87c/src/bin/pgcopydb/ld_apply.c#L1131).

While approaches 1 and 2 are being utilized by the receive process [here to skip](https://github.com/dimitri/pgcopydb/blob/379d242baca0d0fa9bf358d7566bbe74e466e87c/src/bin/pgcopydb/ld_stream.c#L1543-L1547), approach 3 is used by the [apply process](https://github.com/dimitri/pgcopydb/blob/379d242baca0d0fa9bf358d7566bbe74e466e87c/src/bin/pgcopydb/ld_apply.c#L647-L654).

My question is, should we solely rely on the LSN retrieved from the replication origin progress tracker to skip already committed transactions on apply process, and can the receive and transform processes simply act dumb and translate from JSON to SQL?

**Comment by arajkumar** [comment]
>It is expected that resuming logical decoding could send some LSNs that were already processed yes. That's to do with the protocol reporting of writte_lsn, flush_lsn, and apply_lsn. That said, it's easy on the client-side to skip already written LSNs when receiving them again.

@dimitri I think the problem is the order of LSN, we receive it based on it's commit order. Here is a list of LSN which is ordered as it is received.,

```
DC/CC020580
DC/CC0215E8
DC/CC0217F8
DC/CC023BD8
DC/CC01E610
DC/CC01FCF0
DC/CC01FE90
```

When I stop the pgcopydb on `DC/CC01FE90` and resume, we might start receiving data from `DC/CC020580` which we never skip because it is higher than `DC/CC01FE90`.

**Comment by dimitri** [comment]
I see. Well @shubhamdhama did most of the work on resuming logical decoding and interleaved transactions so I would like to have his take on the situation too. What the current code tries to achieve is to continue writing to the JSON file at resume without having to parse it or be smart about it in any way, and also keeping what we already had wrote in the previous round. I suppose that approach is not going to work.

I'm starting to see that logical decoding resume needs to make sense of what's in the JSON file(s) that we have on-disk already and compare to what Postgres is now sending. I still believe we receive the same transactions in the same ordering each time, so it's all about being able to split the existing file(s) and overwrite what we already had with what we are receiving again. 

If we approach this situation as a “cache invalidation” exercise, we might be able to fix the current situation. What do you think @arajkumar?

**Comment by arajkumar** [comment]
@dimitri It makes sense to me. May be I can connect with @shubhamdhama and work with him to find a right fix.

**Comment by shubhamdhama** [comment]
@arajkumar can you please again share the steps to reproduce this including
- Script/queries you are using to write on the source and schema if that's okay.
- How are you using pgcopydb here? Any particular moment you do Ctrl+C.

**Comment by arajkumar** [comment]
@shubhamdhama

> Script/queries you are using to write on the source and schema if that's okay.

I'm generating load using https://github.com/timescale/tsbs. You can follow the below instructions, make sure that you have go toolchain installed and exposed to PATH.

```sh
git clone https://github.com/timescale/tsbs
make tsbs_load

./cmd/tsbs_load load timescaledb --config ./config.yaml

# use the below config.yaml
```

```yaml
# config.yaml
data-source:
  simulator:
    debug: 0
    initial-scale: "0"
    log-interval: 10s
    max-data-points: "0"
    max-metric-count: "1000"
    # max-metric-count: "20000000000"
    # scale: "1"
    scale: "90"
    seed: 100
    timestamp-start: "2016-01-24T00:00:00Z"
    timestamp-end: "2023-01-25T00:00:00Z"
    use-case: iot
  type: SIMULATOR
loader:
  db-specific:
    admin-db-name: <<db>>
    chunk-time: 12h0m0s
    create-metrics-table: true
    field-index: VALUE-TIME
    field-index-count: 0
    force-text-format: false
    host: <<host>>
    in-table-partition-tag: false
    log-batches: true
    partition-index: true
    partitions: 0
    pass: <<password>>
    port: <<port>>
    postgres: sslmode=require
    replication-factor: 0
    time-index: true
    time-partition-index: false
    use-hypertable: false
    use-insert: false
    use-jsonb-tags: false
    user: <<user>>
    write-profile: ""
    write-replication-stats: ""
  runner:
    batch-size: "1200"
    # batch-size: "20000"
    channel-capacity: "0"
    db-name: <<db>>
    do-abort-on-exist: false
    do-create-db: false
    do-load: true
    flow-control: true
    hash-workers: true
    insert-intervals: "1"
    limit: "0"
    reporting-period: 2s
    seed: 100
      # workers: "5"
    workers: "45"
  target: timescaledb
```

Source and Target can be a normal Postgres or TimescaleDB instance. Make sure to replace <<host>>, <<port>>, <password>>,  <<db>> and  <<user>> with appropriate values.

> How are you using pgcopydb here? Any particular moment you do Ctrl+C.

I'm testing only the CDC part like below,
```sh
# clean existing data if any
psql $PGCOPYDB_TARGET_PGURI -c "truncate table readings, tags, diagnostics cascade" && psql $PGCOPYDB_SOURCE_PGURI -c "truncate table readings, tags, diagnostics cascade" && pgcopydb stream cleanup

# start pgcopydb in CDC only mode
pgcopydb follow --plugin wal2json --restart -vv

# you can ctrl+c at any random time and resume pgcopydb
 pgcopydb follow --plugin wal2json --resume -vv

# Repeat resume after Ctrl C at random intervals
```

**Comment by shubhamdhama** [comment]
Thank you for sharing this. Before I try to reproduce, let's try to understand the logs you have already attached. The logs you attached would be useful here. Do you mind sharing `/home/ubuntu/.local/share/pgcopydb/00000002000000B900000042.json` file (and *41.json too, this is not needed but asking just in case)?

**Comment by arajkumar** [comment]
@shubhamdhama , Unfortunately I didn't retain them. I could reproduce it and share new set of files if you want.

**Comment by arajkumar** [comment]
Meanwhile, I was testing an approach which calculates the beginning offset of last LSN from `latest` file and skips all message until the last LSN's insert position. Here is a patch,

https://github.com/arajkumar/pgcopydb/commit/14dc41d821ddab70b769b7e7f58595d4e12e4629.patch

**Comment by arajkumar** [comment]
@shubhamdhama I managed to reproduce the problem again, refer to the attached zip archive.

[Archive.zip](https://github.com/dimitri/pgcopydb/files/12727232/Archive.zip)

**Comment by shubhamdhama** [comment]
Thank you for sharing the logs and CDC files.
```
2023-09-26 11:48:51 161917 SQL    pgsql.c:3715              START_REPLICATION SLOT "pgcopydb" LOGICAL E1/62799C78 ("format-version" '2', "include-xids" 'true', "include-schemas" 'true', "include-transaction" 'true', "include-types" 'true', "filter-tables" 'pgcopydb.*')


$ grep -E "Received action (B|C)" follow.1.2.log | head -n 20
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051499 at LSN E1/627BC350
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051499 at LSN E1/62812FA8
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051501 at LSN E1/626DDCF8
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051501 at LSN E1/6282D300
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051485 at LSN E1/625F1490
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051485 at LSN E1/62831338
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051500 at LSN E1/627D04A0
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051500 at LSN E1/6283B688
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051503 at LSN E1/627983F8
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051503 at LSN E1/62866A28
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051490 at LSN E1/62622490
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051490 at LSN E1/6286D388
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051498 at LSN E1/6271C378
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051498 at LSN E1/62871C18
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051487 at LSN E1/625F9BC0
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051487 at LSN E1/62879160
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051482 at LSN E1/625B9370
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051482 at LSN E1/6289EBF0
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action B for XID 5051496 at LSN E1/6270A160
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051496 at LSN E1/628A5810



Commit for E1/62799C78
2023-09-26 11:48:52 161917 DEBUG  ld_stream.c:698           Received action C for XID 5051503 at LSN E1/62866A28
```

Now I'm on the same page and the fix is what @dimitri  already mentioned
> I'm starting to see that logical decoding resume needs to make sense of what's in the JSON file(s) that we have on-disk already and compare to what Postgres is now sending. I still believe we receive the same transactions in the same ordering each time, so it's all about being able to split the existing file(s) and overwrite what we already had with what we are receiving again.

I'm re-stating the issue here for completeness sake in case someone is not able to follow,
https://github.com/dimitri/pgcopydb/blob/17a317d2647e7e94454e6d85107aa8316a67025e/src/bin/pgcopydb/ld_stream.c#L1531-L1547

When we resume after interruption, we read the latest.json file and make this assumption that when we continue streaming it would stream the last transaction which was partially flushed. But from this issue it's evident that it's not always the case.
In this case in the file `00000003000000E100000062.json`

```
{"action":"I","xid":"5051503","lsn":"E1/62799C78","timestamp":"2023-09-26 11:48:31.318713+0000","message":{"action":"I","xid":5051503,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 07:43:10+00"},{"name":"tags_id","type":"integer","value":902},{"name":"fuel_state","type":"double precision","value":0.6},{"name":"current_load","type":"double precision","value":4974},{"name":"status","type":"double precision","value":0},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"5051503","lsn":"E1/62799C78","timestamp":"2023-09-26 11:48:31.318729+0000","message":{"action":"I","xid":5051503,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 07:43:10+00"},{"name":"tags_id","type":"integer","value":14517},{"name":"fuel_state","type":"double precision","value":0.6},{"name":"current_load","type":"double precision","value":4974},{"name":"status","type":"double precision","value":4},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"5051503","lsn":"E1/62799C78","timestamp":"2023-09-26 11:48:31.318739+0000","message":{"action":"I","xid":5051503,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 07:43:20+00"},{"name":"tags_id","type":"integer","value":902},{"name":"fuel_state","type":"double precision","value":0.6},{"name":"current_load","type":"double precision","value":2971},{"name":"status","type":"double precision","value":0},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"K","lsn":"E1/62799C78","timestamp":"2023-09-26 11:48:31.318739+0000"}
```
there was an interruption after these lines and when restarted the follow, it picked `startpos` as `E1/62799C78`. Exepectation was it will continue streaming from `E1/627983F8`. Instead, it started from few transactions back, I suppose the first "start of a transaction (B)" before the `E1/62799C78`. 

```

**Comment by shubhamdhama** [comment]
I propose the following solution for dealing with partially flushed transaction by ignore already commited transactions, reach the partially written transaction and continue the existing process. 
1. Get the last commit LSN that was flushed earlier 
    - There could be multiple ways to achieve this. We just need to find the most effecient one that involves minimal reading of previous files for the transaction that spans multiple files. How about start from a file whose name is calculated from replication origin tracking LSN or another relevant LSN. Keep reading messages and note the last seen commit LSN. Alternatively, we might consider designing a new device within the "stream process" specifically to monitor the last commit, which would be useful during the "recovery process".
2. As streaming resumes, create a temporary/staging file (or possibley multiple files, depending upon naming) where we would keep writing messages to look for their commit LSNs and hence avoiding keeping everything in memory
    - discard the incoming transactions whose commit LSN is less than or equal to the lastStreamedCommitLSN.
    - if commit LSN is greater than lastStreamedCommitLSN then merge it with the main json file after ignoring the messages already written (existing process)
 
A slight improvement here would be to delete all the lines after `lastStreamedCommitLSN` in "1". This would make merge easier in "2".

Please share your thoughts. I could be mistaken in the root cause. Or perhaps a simpler solution than this proposal can come easily.

**Comment by arajkumar** [comment]
> Meanwhile, I was testing an approach which calculates the beginning offset of last LSN from `latest` file and skips all message until the last LSN's insert position. Here is a patch,
> 
> https://github.com/arajkumar/pgcopydb/commit/14dc41d821ddab70b769b7e7f58595d4e12e4629.patch

What do you think of this approach? Here is a visual representation of the same,
![image](https://github.com/dimitri/pgcopydb/assets/3874763/1b7e2936-6312-46b5-809b-49c3229e606c)

The idea is, while resuming, 
1. find the last LSN from the latest.json
2. find line offset between last message of last LSN and first message of last LSN
3. Ignore all messages until last LSN
4. Start counting lines and ignore until the count reaches the offset (calculated in step 2)
5. Resume appending to latest.json file

**Comment by shubhamdhama** [comment]
@dimitri, I'm again out of context what we finalized on this, can you please put it here so that @arajkumar can go ahead on this?
I think it was simply finding the position where the restarted streaming should start writing and overwrite the data from that position.

**Comment by dimitri** [comment]
I agree with you @shubhamdhama ; I think what should happen here is “cache invalidation” where we rewrite data that we already received (partially) in the previous rounds. 

Comment https://github.com/dimitri/pgcopydb/issues/471#issuecomment-1741988692 by @arajkumar  seems to be more in the spirit of trusting the local cache and resuming writes once we are sync'ed again with the source server, I think that's more prone to errors, given the issues we are trying to fix at the moment.

**Comment by arajkumar** [comment]
> I think it was simply finding the position where the restarted streaming should start writing and overwrite the data from that position.

@dimitri  I think this would work as long as the restarted streaming LSN falls with the latest WAL segment boundary, I'm worried about what happens if it goes back to the previous WAL segment.? Maybe, we keep writing to the latest WAL segment file from the beginning?

**Comment by dimitri** [comment]
I'm okay with having “holes” in the first round of implementation, that is, error out in that case for the moment, so that we have a solid code base and review that later. Again, I'm starting to think that our WAL file naming strategy is going to have to go back to the design board, and that's one of the problems that we collect and need the new design to solve elegantly.

**Comment by arajkumar** [comment]
@dimitri @shubhamdhama I want to add another option which I'm considering; The idea is to append a "ROLLBACK" message incase if we encounter a partial transaction(i.e. the last message is not a COMMIT) in the latest file while resuming. This would avoid us from handing a txn without a proper ending(i.e. COMMIT).

For example, consider a latest(.json) which contains a partial txn(i.e. with a COMMIT) due to graceful termination(either Ctrl C or soft kill from prefetch to replay switchover)

```.json
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597955+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5769},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":5},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597963+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5770},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":1},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597986+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5774},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":3},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"K","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597986+0000"}
```

When resuming this, instead of doing error prune calculation of finding the right place to resume we can simply add a ROLLBACK message into the end if last logical message received from the source is not COMMIT. This will help us to easily handle this in transform and apply without managing a transaction state. 

The modified latest(.json) file will look like below,

```.json
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597955+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5769},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":5},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597963+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5770},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":1},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"I","xid":"9168494","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597986+0000","message":{"action":"I","xid":9168494,"schema":"public","table":"diagnostics","columns":[{"name":"time","type":"timestamp with time zone","value":"2016-01-24 01:30:20+00"},{"name":"tags_id","type":"integer","value":5774},{"name":"fuel_state","type":"double precision","value":0.7},{"name":"current_load","type":"double precision","value":2395},{"name":"status","type":"double precision","value":3},{"name":"additional_tags","type":"jsonb","value":null}]}}
{"action":"K","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597986+0000"}
{"action":"R","lsn":"186/97F8BDD0","timestamp":"2023-11-06 08:34:44.597986+0000"}
```

**Comment by dimitri** [comment]
That would work (paying attention to the last message received being in a transaction or a standalone message, of course, and KEEPALIVE could be found either way) ; but when debugging having the same transaction added multiple times to the same file is going to make things harder to reason about I believe.

**Comment by arajkumar** [comment]
@dimitri Identifying the right place to start from the latest file is bit challenging with our logic to always append to the latest segment.
https://github.com/dimitri/pgcopydb/blob/3fda744f77512252caee9037be96e884b9fcdbc7/src/bin/pgcopydb/ld_stream.c#L1065-L1082

Consider the following scenario, where the txn 2 spawns across 2 segments(B.json & C.json) and we gracefully exit right after writing the first message. When we resume, the we would start receiving the full txn body, which actually starts in A.json. Now we have 3 different files in order to resume correctly.

#### A.json
```.json
{"action": "B", "xid": 1, "lsn": "A", ...}
....
{"action": "I", "xid": 1, "lsn": "B", ...}
{"action": "X", "lsn": "B", ...}

```

#### B.json
```.json
{"action": "C", xid: 1, "lsn": "B", ...}
...
...
{"action": "B", "xid": 2, "lsn": "A", ...}
{"action": "I", "xid": 2, "lsn": "A", ...}
...
{"action": "I", "xid": 2, "lsn": "B" ... }
{"action": "I", "xid": 2, "lsn": "B" ... }
...
{"action": "I", "xid": 2, "lsn": "C" ... }
{"action": "X", "xid": 2, "lsn": "C" ... }
``` 

#### latest -> C.json
```.json
{"action": "I", "xid": 2, "lsn": "C" ... }
{"action": "K", "lsn": "C" }
``` 

I think we have to take the tradeoff in debugging, it might look incorrect to have 2 transactions, but having to deal with multiple segments would be error prune too. WDYT?

**Comment by dimitri** [comment]
The main problem is that we use the WAL filenames extracted from LSN positions, as Postgres does for its physical WAL files; but we're using logical decoding where we receive transactions in full and messages on these transactions can be spread in many ways.

In your case though, I believe it's a matter of how we initialize maxWrittenLSN, which at the moment is set to `startpos` in cases where we might have to set it to the latest LSN read in the latest JSON file. Because then we're using it as a low watermark, IIUC, and we just skip messages that are sent and found before what we have already written to disk.

The alternative approach that I think we should consider is to overwrite all messages that we have already written to disk when we receive them again. For that we need to compute the JSON filename from the startpos received from the Postgres replication protocol, open the file and truncate it (keep only up to the startpos), and start writing data again. Later files that already exist should also be renamed before we start writing.

Do you think that approach is possible to implement?

**Comment by arajkumar** [comment]
>In your case though, I believe it's a matter of how we initialize maxWrittenLSN, which at the moment is set to startpos in cases where we might have to set it to the latest LSN read in the latest JSON file. Because then we're using it as a low watermark, IIUC, and we just skip messages that are sent and found before what we have already written to disk.

startpos is actually assigned based on the latest LSN from the latest JSON. https://github.com/dimitri/pgcopydb/blob/a6a6e3c02283065737b959086efcd1c645735c23/src/bin/pgcopydb/ld_stream.c#L690-L692

**Comment by arajkumar** [comment]
@dimitri I've implemented the following suggestion,
>The alternative approach that I think we should consider is to overwrite all messages that we have already written to disk when we receive them again. For that we need to compute the JSON filename from the startpos received from the Postgres replication protocol, open the file and truncate it (keep only up to the startpos), and start writing data again. Later files that already exist should also be renamed before we start writing.

https://github.com/dimitri/pgcopydb/pull/536

I'm still validating it, hence keeping the PR as draft.

**Comment by arajkumar** [comment]
Fixed by https://github.com/dimitri/pgcopydb/issues/544.
